conn.go

```diff
@@ -821,21 +747,56 @@ func (c *conn) exchangeProtoHeader(pID protoID) (stateFunc, error) {
 
 // readProtoHeader reads a protocol header packet from c.rxProto.
 func (c *conn) readProtoHeader() (protoHeader, error) {
-	var deadline <-chan time.Time
-	if c.connectTimeout != 0 {
-		deadline = time.After(c.connectTimeout)
+	// only read from the network once our buffer has been exhausted.
+	// TODO: this preserves existing behavior as some tests rely on this
+	// implementation detail (it lets you replay a stream of bytes). we
+	// might want to consider removing this and fixing the tests as the
+	// protocol doesn't actually work this way.
+	if c.rxBuf.Len() == 0 {
```

Comment:
Is it possible to have _some_ bytes in rxBuf but just not enough to be a complete protocol header? If so, we should go in here if c.rxBuf.Len() < protoHeaderSize?

Reply:
We check `c.rxBuf.Len()` for the proper length in the `for` loop below and will keep reading as required.